StarRating/StarRating/CustomControl.swift

```diff
@@ -0,0 +1,143 @@
+            newLabel.text = "☆"
+            newLabel.textAlignment = .center
+            if i == 1 {
+                newLabel.textColor = componentActiveColor
+            } else {
+                newLabel.textColor = componentInactiveColor
+            }
+            self.addSubview(newLabel)
+            labels.append(newLabel)
+        }
+    }
+}
+
+//MARK: - Touch Handler
+extension CustomControl {
+    override func beginTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
+        updateValue(at: touch)
+        
+        sendActions(for: [.touchDown, .valueChanged])
+        
+        return true
+    }
+    
+    override func continueTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
+        if self.bounds.contains(touch.location(in: self)) {
+            updateValue(at: touch)
+            
+            sendActions(for: [.touchDragInside, .valueChanged])
+        } else {
+            sendActions(for: [.touchDragOutside])
+        }
+        return true
+    }
+    
+    override func endTracking(_ touch: UITouch?, with event: UIEvent?) {
+        
+        defer {
+            super.endTracking(touch, with: event)
+        }
+        guard let touch = touch else { return }
+        
+        if bounds.contains(touch.location(in: self)) {
+            updateValue(at: touch)
+            
+            sendActions(for: [.touchUpInside, .valueChanged])
+        } else {
+            sendActions(for: [.touchUpOutside])
+        }
+        
+    }
+    
+    override func cancelTracking(with event: UIEvent?) {
+        sendActions(for: [.touchCancel])
+    }
+    
+    func updateValue(at touch: UITouch) {
+        print(touch.location(in: self))
+        
+        for label in labels {
+            
+            if label.frame.contains(touch.location(in: self)) && value != label.tag {
+                oldValue = value
+                value = label.tag
+                if value > oldValue {
+                    for i in 1...label.tag {
+                        if let updateLabel = self.viewWithTag(i) as? UILabel {
+                            updateLabel.textColor = componentActiveColor
+                            updateLabel.performFlare()
+                        }
+                    }
+                }
+                if value < oldValue {
+                    for i in stride(from: labels.count, to: label.tag, by: -1) {
+                        if let updateLabel = self.viewWithTag(i) as? UILabel {
+                            updateLabel.textColor = componentInactiveColor
+                        }
+                    }
+                    for j in 1...label.tag {
+                        if let updateLabel = self.viewWithTag(j) as? UILabel {
+                            updateLabel.performFlare()
+                        }
+                    }
+                }
+                
+            }
+        }
+    }
+}
+extension UIView {
+    // "Flare view" animation sequence
+    func performFlare() {
```

Comment:
This is great, but I would suggest that you would use closures instead to make code more readable.

```
let unflare = {
 transform = .identity
}
```